src/Microsoft.TemplateEngine.Cli/TemplateResolution/TemplateResolutionResult.cs

```diff
@@ -5,246 +5,290 @@
-                {
-                    _bestTemplateMatchList = BaseGetBestTemplateMatchList(ignoreDefaultLanguageFiltering);
-                }
+            IEnumerable<ITemplateMatchInfo> highestPrecedenceTemplates = UnambiguousTemplateGroup.GetHighestPrecedenceInvokableTemplates(!_hasUserInputLanguage);
+            IEnumerable<string> templateLanguages = highestPrecedenceTemplates.Select(t => t.Info.GetLanguage()).Distinct(StringComparer.OrdinalIgnoreCase);
 
-                return _bestTemplateMatchList;
-            }
-            else
+            if (templateLanguages.Count() > 1)
             {
-                if (_bestTemplateMatchListIgnoringDefaultLanguageFiltering == null)
-                {
-                    _bestTemplateMatchListIgnoringDefaultLanguageFiltering = BaseGetBestTemplateMatchList(ignoreDefaultLanguageFiltering);
-                }
-
-                return _bestTemplateMatchListIgnoringDefaultLanguageFiltering;
+                _singularInvokableMatchStatus = Status.AmbiguousLanguageChoice;
+                return;
             }
+            _singularInvokableMatchStatus = Status.AmbiguousTemplateChoice;
         }
 
-        // The core matched templates should not need additioanl default language filtering.
-        // The default language dispositions are stored in a different place than the other dispositions,
-        // and are not considered for most match filtering.
-        private IReadOnlyList<ITemplateMatchInfo> BaseGetBestTemplateMatchList(bool ignoreDefaultLanguageFiltering)
+        private void EvaluateUnambiguousTemplateGroup()
         {
-            IReadOnlyList<ITemplateMatchInfo> templateList;
-            if (TryGetUnambiguousTemplateGroupToUse(out templateList, ignoreDefaultLanguageFiltering))
-            {
-                return templateList;
-            }
-            else if (!string.IsNullOrEmpty(_templateName) && TryGetAllInvokableTemplates(out templateList))
-            {
-                return templateList;
-            }
-            else if (TryGetCoreMatchedTemplatesWithDisposition(x => x.IsMatch, out templateList))
-            {
-                return templateList;
-            }
-            else if (TryGetCoreMatchedTemplatesWithDisposition(x => x.IsMatchExceptContext(), out templateList))
-            {
-                return templateList;
-            }
-            else if (TryGetCoreMatchedTemplatesWithDisposition(x => x.IsPartialMatch, out templateList))
+            if (TemplateGroups.Count == 0)
             {
-                return templateList;
+                _unambigiousTemplateGroupStatus = UnambiguousTemplateGroupStatus.NoMatch;
+                return;
             }
-            else if (TryGetCoreMatchedTemplatesWithDisposition(x => x.IsPartialMatchExceptContext(), out templateList))
+            if (TemplateGroups.Count == 1)
             {
-                return templateList;
+                _unambiguousTemplateGroup = TemplateGroups.First();
```

Comment:
nit: change `TemplateGroups.First()` to `TemplateGroups.Single()` for better readability, my brain stopped for second, why do we just pick up 1st group, and then I saw `if` before...